qiskit_experiments/framework/experiment_data.py

```diff
@@ -767,6 +770,66 @@ def add_data(
     ) -> None:
         """Add experiment data.
 
+        Args:
+            data: Experiment data to add. Several types are accepted for convenience:
+
+                * Result: Add data from this ``Result`` object.
+                * List[Result]: Add data from the ``Result`` objects.
+                * Dict: Add this data.
+                * List[Dict]: Add this list of data.
+
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+
+        if any(not future.done() for future in self._analysis_futures.values()):
+            LOG.warning(
+                "Not all analysis has finished running. Adding new data may "
+                "create unexpected analysis results."
+            )
+        if not isinstance(data, list):
+            data = [data]
+
+        # Directly add non-job data
+        with self._result_data.lock:
+            tmp_exp_data = ExperimentData()
+            composite_flag = False
+            experiment_seperator = {}
+            for datum in data:
+                if isinstance(datum, dict):
+                    if "composite_metadata" in datum["metadata"]:
+                        composite_flag = True
+                        marginalized_data = self._marginalized_component_data([datum])
```

Comment:
Since this call is always with a single element, it's much simpler to assume `Dict` as an input instead of `List[Dict]`.